### PR TITLE
Improved scheduling page

### PR DESCRIPTION
**Pull request by TimGriffioen**

### CODE REVIEW

File: src/lib/components/scheduling/SongAvailibilityTimeline.svelte [hunk 1]
```diff
@@ -81,17 +81,15 @@
...
-		<Column class="somebody-unavailable">Somebody is unavailable</Column>
-	{:else if timeslotPossible}
+	{#if timeslotPossible}
 		{#if minStartTime > 0}
 			<Column
 				class="red"
 				style={`max-width: ${(minStartTime / totalMinutes) * 100}% !important;`}
 			/>
 		{/if}
 		<Column
```
Comment (BMN-Comm) [inline src/lib/components/scheduling/SongAvailibilityTimeline.svelte:91]: Orange now shows if some people are completely unavailable, but don't you also want it to show if for a certain period 1 person is not available at that time, even though they may be later?

File: src/lib/firebase/client/firestore/musicians.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { db } from '$lib/firebase/client/firebase'
 import type { user } from '$lib/types/domain/user'
 import { QueryWhereInBatched } from '$lib/util/queryWhereIn'
-import { collection, collectionGroup, doc, getDoc } from 'firebase/firestore'
+import { collection, collectionGroup, doc, getDoc, query } from 'firebase/firestore'
```
Comment (BMN-Comm) [inline src/lib/firebase/client/firestore/musicians.ts:4]: heck? (it is not used right?)

File: src/routes/committee/rehearsals/+page.svelte [hunk 3]
```diff
@@ -61,18 +63,72 @@
 		date = ''
 		startTime = '18:00'
 		endTime = '21:00'
-		openModal = false
+
+		invalidateAll()
```
Comment (BMN-Comm) [inline src/routes/committee/rehearsals/+page.svelte:67]: don't forget to close the add modal

File: src/routes/committee/rehearsals/+page.svelte [hunk 4]
```diff
@@ -61,18 +63,72 @@
...
+	/** Edit a rehearsal in the database */
+	async function editRehearsal() {
+		// First delete the old document
+		const docRef = doc(db, 'rehearsals', data.rehearsals[selectedRehearsalId].id)
+		await deleteDoc(docRef)
+
+		// Then add a new rehearsal with the new data
+		let dateSplit = date.split('/')
+		let sTimeSplit = startTime.split(':')
+		let eTimeSplit = endTime.split(':')
```
Comment (BMN-Comm) [inline src/routes/committee/rehearsals/+page.svelte:79]: I'm a bit out of practice with firebase, but can't you modify a document directly?

File: src/routes/committee/rehearsals/+page.svelte [hunk 5]
```diff
@@ -61,18 +63,72 @@
...
 
 	/** Remove a rehearsal from the database */
 	async function removeRehearsal() {
-		const docRef = doc(db, 'rehearsals', data.rehearsals[selectedRehearsal].id)
+		const docRef = doc(db, 'rehearsals', data.rehearsals[selectedRehearsalId].id)
+
 		await deleteDoc(docRef)
 
 		invalidateAll()
 	}
```
Comment (BMN-Comm) [inline src/routes/committee/rehearsals/+page.svelte:111]: don't you need to close them here too?

File: src/routes/committee/rehearsals/+page.svelte [hunk 6]
```diff
@@ -61,18 +63,72 @@
...
+		
+		const endTimeDate = rehearsal.endTime.toDate()
+		const endTimeHours = endTimeDate.getHours()
+		const endTimeMinutes = endTimeDate.getMinutes()
+		endTime = endTimeHours + ':' + ("0" + endTimeMinutes).slice(-2)
+
+		const dateDay = startTimeDate.getDate()
+		const dateMonth = startTimeDate.getMonth() + 1
+		const dateYear = startTimeDate.getFullYear()
+		date = dateDay + '/' + dateMonth + '/' + dateYear
```
Comment (BMN-Comm) [inline src/routes/committee/rehearsals/+page.svelte:130]: all consts are used once here, I'd just inline them

File: src/routes/committee/rehearsals/+page.svelte (hunk 4 above)
Comment (Harrimaga) [reply src/routes/committee/rehearsals/+page.svelte:79]: Yeah you can update a doc

Comment (Harrimaga) [reply src/routes/committee/rehearsals/+page.svelte:79]: https://firebase.google.com/docs/firestore/manage-data/add-data#update-data

File: src/routes/committee/rehearsals/+page.svelte [hunk 7]
```diff
@@ -186,19 +253,66 @@
...
+	bind:open={openEditModal}
+	on:click:button--primary={() => {
+		editRehearsal()
+		openEditModal = false
+	}}
+	on:click:button--secondary={() => {
+		openEditModal = false
+	}}
+>
+<div class="modal">
```
Comment (BMN-Comm) [inline src/routes/committee/rehearsals/+page.svelte:270]: does your prettier work? I feel like this should have a tab in front of it

File: src/routes/committee/rehearsals/[rehearsalId]/schedule/+page.svelte [hunk 8]
```diff
@@ -144,6 +144,10 @@
 							items={songs.map((s) => ({ id: s.id, text: s.name }))}
 							bind:selectedId={songId}
 							required
+							shouldFilterItem={(item, value) => {
+								if (!value) return true;
+								return item.text.toLowerCase().includes(value.toLowerCase())
+							}}
```
Comment (BMN-Comm) [inline src/routes/committee/rehearsals/[rehearsalId]/schedule/+page.svelte:150]:
```suggestion
							shouldFilterItem={(item, value) =>
								value ? item.text.toLowerCase().includes(value.toLowerCase()) : true;
							}
```

File: src/routes/committee/rehearsals/[rehearsalId]/availability/+page.svelte [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+	const usersWithUndefinedAvailabilities = data.userAvailabilities.filter((x) => !x.availability)
+	const usersWithDefinedAvailabilities = data.userAvailabilities.filter((x) => !!x.availability)
+	const unavailableUsers = usersWithDefinedAvailabilities.filter((x) => !x.availability!.available).map(
+		(user) => {
+			return {
+				name: user.name,
+				reason: user.availability!.reason ? user.availability!.reason : 'No reason given'
+			}
+		}
```
Comment (BMN-Comm) [inline src/routes/committee/rehearsals/[rehearsalId]/availability/+page.svelte:25]:
```suggestion
		(user) => ({
			name: user.name,
			reason: user.availability!.reason ? user.availability!.reason : 'No reason given'
		})
```
		also change in const below this one

File: src/lib/components/scheduling/SongAvailibilityTimeline.svelte (hunk 1 above)
Comment (TimGriffioen) [reply src/lib/components/scheduling/SongAvailibilityTimeline.svelte:91]: Yes, I have to look into this still. For now just accept it as it is. Can't fix it in the time I have left today, so I will open another Issue for it

File: src/lib/components/scheduling/ScheduleTimeline.svelte [hunk 10]
```diff
@@ -82,12 +80,6 @@
 </Row>
 
 <style>
-	.vertical-text {
-		writing-mode: vertical-rl;
-		text-orientation: mixed;
-		margin-left: -10px;
-	}
-
```
Comment (RienkF) [inline src/lib/components/scheduling/ScheduleTimeline.svelte:90]: why no vertical?

File: src/lib/components/scheduling/SongAvailibilityTimeline.svelte [hunk 11]
```diff
@@ -159,6 +161,10 @@
 		background-color: #ef0612;
 	}
```
Comment (RienkF) [inline src/lib/components/scheduling/SongAvailibilityTimeline.svelte:162]: is red still needed?

File: src/lib/components/scheduling/TimelineBlock.svelte [hunk 12]
```diff
@@ -1,16 +1,13 @@
...
 
 	export let relativeWidth: number
 	export let song: song | undefined = undefined
 	export let deleteSong: (() => void) | undefined = undefined
-
-	$: randomColor = song ? stringToColour(song.id) : '#808080'
 </script>
 
-<div class="song-block" style={`width: ${relativeWidth}%; background-color: ${randomColor}`}>
+<div class="song-block" style="--relative-width: {relativeWidth}%;">
```
Comment (RienkF) [inline src/lib/components/scheduling/TimelineBlock.svelte:10]: But, but, mah colour...

File: src/routes/committee/rehearsals/+page.svelte [hunk 13]
```diff
@@ -58,21 +60,62 @@
...
+		// Then add a new rehearsal with the new data
+		console.log(date, startTime, endTime)
+
+		let dateSplit = date.split('/')
+		let sTimeSplit = startTime.split(':')
+		let eTimeSplit = endTime.split(':')
+
+		console.log(dateSplit, sTimeSplit, eTimeSplit)
+
+		await updateDoc(docRef, {
```
Comment (RienkF) [inline src/routes/committee/rehearsals/+page.svelte:85]: Remove logs and update comment

**Review by RienkF — APPROVED** [review]

File: src/lib/components/scheduling/ScheduleTimeline.svelte (hunk 10 above)
Comment (TimGriffioen) [reply src/lib/components/scheduling/ScheduleTimeline.svelte:90]: Takes up more space and imo is less readable

File: src/lib/components/scheduling/SongAvailibilityTimeline.svelte (hunk 11 above)
Comment (TimGriffioen) [reply src/lib/components/scheduling/SongAvailibilityTimeline.svelte:162]: Yes, currently for when people are partly unavailable, that part is red, but I will improve it in a future issue.

File: src/lib/components/scheduling/TimelineBlock.svelte (hunk 12 above)
Comment (TimGriffioen) [reply src/lib/components/scheduling/TimelineBlock.svelte:10]: :)

File: src/routes/committee/rehearsals/+page.svelte (hunk 13 above)
Comment (TimGriffioen) [reply src/routes/committee/rehearsals/+page.svelte:85]: Done

**Pull request merged**